## 1. Loading the Data:

In [1]:
# import libraries that will be used
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
# read csv file
df = pd.read_csv(r"C:\Users\adiya\Documents\Uni\Data Science\1SA-Final-Project\hotel_bookings.csv")

In [3]:
# display the first 5 rows for a quick look
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

## 2. Understanding the Data:

In [4]:
# check the shape of the DataFrame (rows, columns)
# understand the amount of data
df.shape

(119390, 32)

In [5]:
# description of data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

> In a first observation it is clear some features (for example "company",
> and "agent") consist of many null values.

In [6]:
# summary of the numerical attributes
df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

> ### Features in the DataFrame:
> 0. hotel: Resort Hotel or City Hotel
> 1. is_cancelled: Value indicating if the booking was canceled (1) or not (0)
> 2. lead_time: Number of days between the booking date to the arrival date
> 3. arrival_date_year: Year of arrival date
> 4. arrival_date_month: Month  of arrival date
> 5. arrival_date_week: Week number according to year of arrival
> 6. arrival_date_day: Day (in numbers) of arrival date
> 7. stays_in_weekend_nights: Number of weekend nights (Saturday or Sunday) booked for accommodation
> 8. stays_in_week_nights: Number of week nights (Monday to Friday) booked for accommodation
> 9. adults: Number of adults
> 10. children: Number of children
> 11. babies: Number of babies
> 12. meal: Type of meal booked
> 13. country: Country of origin
> 14. market_segment: Market segment designation (TA- Travel Agents, TO- Tour Operators)
> 15. distribution_channel: Booking distribution channel (TA- Travel Agents, T0- Tour Operators)
> 16. is_repeated_guest: Value indication if booking name was from a repeated guest (1) or not (0)
> 17. previous_cancellations: Number of previous bookings that were cancelled by the customer (prior to current booking)
> 18. previous_bookings_not_canceled: Number of previous bookings not cancelled prior to current booking
> 19. reserved_room_type: Code of room type reserved
> 20. assigned_room_type: Code for the type of room assigned to the booking
> 21. booking_changes: Number of changes made to the booking
> 22. deposit_type: Indication on if the customer made a deposit to guarantee the booking
> 23. agent: ID of the travel agency that made the booking
> 24. company: ID of the company or organization that made the booking or is responsible for payment
> 25. days_in_waiting_list: Number of days booking was in the waiting list until it was confirmed to the customer
> 26. customer_type: Type of booking
> 27. adr: Average Daily Rate (the sum of transactions divided by the number of nights stayed)
> 28. required_car_parking_spaces: Number of car parking spaces requested
> 29. total_of_special_requests: Number of special requests made by the customer
> 30. reservation_status: Last reservation status (Canceled, Check-Out, No-Show)
> 31. reservation_status_date: Date at which the last status was set
>

In [7]:
# further examination of null values
# the methods below calculate the number of missing values
missing_values = df.isna().sum()
missing_values = missing_values[missing_values != 0]
missing_values

children         4
country        488
agent        16340
company     112593
dtype: int64

> The "company" and "agent" columns have a large amount of null values.
> Since they are irrelevant for research they can be removed.

In [8]:
# dropping "company" and "agent" columns
df.drop(["company", "agent"], axis=1, inplace=True)

> The children column has only 4 missing values. These can be replaced with 0, suggesting
> there were no kids during the stay.

In [9]:
# replacing null values in 'children' column with 0
df.fillna({"children": 0}, inplace=True)
# examining changes
df['children'].isnull().sum()

0

In [10]:
df["country"].value_counts()

PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
       ...  
NAM        1
NIC        1
UMI        1
DJI        1
NCL        1
Name: country, Length: 177, dtype: int64

> As shown above, there are 177 different countries. This indicates that the data is representative
> worldwide, contrary to a specific region.
> A quick assessment will portray a better perspective on how to treat the missing data.

In [11]:
num1 = (df["country"].value_counts() <= 1).sum()
num10 = (df["country"].value_counts() <= 10).sum()
num50 = (df["country"].value_counts() <= 50).sum()
num100 = (df["country"].value_counts() <= 100).sum()
num1000 = (df["country"].value_counts() <= 1000).sum()

print("Number of countries with one or less instances:", num1,
      "\nNumber of countries with 10 or less instances:", num10,
      "\nNumber of countries with 50 or less instances:", num50,
      "\nNumber of countries with 100 or less instances:", num100,
      "\nNumber of countries with 1000 or less instances:", num1000)

Number of countries with one or less instances: 30 
Number of countries with 10 or less instances: 84 
Number of countries with 50 or less instances: 120 
Number of countries with 100 or less instances: 140 
Number of countries with 1000 or less instances: 162


> More than half of the instances have 50 or fewer observations in the DataFrame.
> Meanwhile, there are few countries with thousands. This shows that the countries
> are not the best option to represent results from the data since they are
> unevenly divided. For all the reasons stated, this column is unnecessary for further
> analysis.

In [12]:
# dropping "country" column
df.drop("country", axis=1, inplace=True)

> #### Quick look after changes:

In [13]:
# examining the changes
df.head(3)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
0                        0                     0       2  ...   
1                        0                     0       2  ...   
2                        0                     1       1  ...   

   assigned_room_type  booking_changes deposit_type days_in_waiting_list  \
0                   C                3   No Deposit                    0   
1                   C                4   No Deposit                    0   
2                   C                0   No Deposit                    0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0     Transient   0.0                            0                          0   
1     Transient   0.0                            0                          0   
2     Transient  75.0                            0                          0   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  

[3 rows x 29 columns]

In [14]:
# checking if there are remaining null values
df.isna().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status          

In [15]:
# number of columns changed to 29
# number of instances is the same
df.shape

(119390, 29)